### PR TITLE
cnap/core:add qat_zip to compress/decompress the frame

### LINKED ISSUE #91: Add QATzip tools to accelerate compression/decompression of frames.
When compressing/decompressing frames, can enable Intel Accelerator to accelerate the operation.

### PR DESCRIPTION
**Pull request by Yanbo0101**
This PR is to fix #91, and it depends on the Intel QAT driver and QATzip.

For usages examples:

#initialization
qat = QATFrameZip() 

#compression
compressed_data = qat.compress(frame_data)

#decompression
decompressed_data = qat.decompress(compressed_data)

#teardown&close
qat.teardown_session()
qat.close_session()

### CODE REVIEW

File: cnap/core/frame.py [hunk 1]
```diff
@@ -58,6 +58,46 @@ def decrypt(self) -> None:
...
+    def compress(self, src: bytes) -> bytes:
+        """Compress the frame.
+
+        The method is to compress the frame to protect the privacy of the frame.
+
+        Args:
+            src (bytes): The frame need to compressed.
+        
+        Return:
+            bytes: bytes: Compressed Frame.
```
Comment (Hulongyin) [inline cnap/core/frame.py]: bytes:

File: cnap/core/frame.py [hunk 2]
```diff
@@ -58,6 +58,46 @@ def decrypt(self) -> None:
...
+    def decompress(self, src: bytes) -> bytes:
+        """Decompress the frame.
+        
+        The method is to decompress the frame to protect the privacy of the frame.
+
+        Args:
+            src (bytes): The frame need to decompressed.
+        
+        Return:
+            bytes: bytes: Deompressed Frame.
```
Comment (Hulongyin) [inline cnap/core/frame.py]: bytes:

File: cnap/core/qat_zip.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
+"""A QAT params module.
+
+This module provides the relevant parameters required to call QAT.
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: The description of this module is the same as `qat_params.py`

File: cnap/core/qat_zip.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    QzPollingMode: An enum for the type of polling mode supported by QATzip.
+"""
+import ctypes
+import logging
+import time
+from core.frame import FrameZipBase
+from core.qat_params import QzSession, QzSessionParamsDeflate
+
+LOG = logging.getLogger(__name__)
+qziplib = ctypes.cdll.LoadLibrary("/usr/local/lib/libqatzip.so")
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Can this path be configurable instead of hardcode?

File: cnap/core/qat_zip.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        Args:
+            src (bytes): Frame that need to be compressed.
+
+        Returns:
+            bytes: Compressed Frame.
+        """
+        dst_sz = qziplib.qzMaxCompressedLength(len(src), ctypes.byref(self._session))
+        if dst_sz == 0:
+            LOG.error("compression integer overflow happens.")
+            return None
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Raise the error may be better.

File: cnap/core/qat_zip.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        if dst_sz == 34:
+            LOG.error("compression input length is 0.")
+            return None
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Same.

File: cnap/core/qat_zip.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        start_time = time.time()
+
+        ret = qziplib.qzCompress(
+            ctypes.byref(self._session),
+            src_point, ctypes.byref(src_len),
+            dst_point, ctypes.byref(dst_len),
+            LAST)
+        if ret != 0:
+            LOG.error("Failed to compress: %s.", ret)
+            return None
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Same.

File: cnap/core/qat_zip.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+        start_time = time.time()
+
+        ret = qziplib.qzDecompress(
+            ctypes.byref(self._session),
+            src_point, ctypes.byref(src_len),
+            dst_point, ctypes.byref(dst_len))
+        if ret != 0:
+            LOG.error("Failed to decompress: %s.", ret)
+            return None
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Same.

File: cnap/core/qat_zip.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        self._deflate_params = deflate_params
+        self.init_session()
+        self.set_session()
+
+    def init_session(self) :
+        """Initialize QAT hardware."""
+        ret = qziplib.qzInit(ctypes.byref(self._session), SW_BACKUP)
+        if ret != 0:
+            LOG.error("Failed to initialize the QATzip session: %s", ret)
+            raise NotImplementedError(f"Initialize QATzip session failed with status:{ret}.")
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Is it appropriate to use `NotImplementedError` here? And the raises should also be added to docstring.

File: cnap/core/qat_zip.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        LOG.info("Initialize QAT hardware successful.")
+
+    def set_session(self) :
+        """Initialize a QATzip session with deflate parameters."""
+        ret = qziplib.qzSetupSessionDeflate(
+            ctypes.byref(self._session),
+            ctypes.byref(self._deflate_params))
+        if ret != 0:
+            LOG.error("Failed to setup the QATzip session: %s.", ret)
+            raise NotImplementedError(f"Set QATzip session failed with status: {ret}.")
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Same.

File: cnap/core/qat_zip.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    def teardown_session(self) :
+        """Uninitialize a QATzip session.
+
+        This method disconnects a session from a hardware instance and deallocates buffers. 
+        If no session has been initialized, then no action will take place.
+        """
+        ret = qziplib.qzTeardownSession(ctypes.byref(self._session))
+        if ret != 0:
+            LOG.error("Failed to teardown the QATzip session: %s.", ret)
+
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Do we need to raise error here?

File: cnap/core/qat_zip.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    def close_session(self) -> None:
+        """Terminates a QATzip session.
+        
+        This method closes the connection with QAT.
+        """
+        ret = qziplib.qzClose(ctypes.byref(self._session))
+        if ret != 0:
+            LOG.error("Failed to close the QATzip session: %s.", ret)
+
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: Do we need to raise error here?

File: cnap/core/frame.py [hunk 13]
```diff
@@ -58,6 +58,46 @@ def decrypt(self) -> None:
...
+
+    This class serves as a blueprint for subclass that need to implement `compress`
+    and `decompress` method for different accelerators.
+    """
+
+    @abstractmethod
+    def compress(self, src: bytes) -> bytes:
+        """Compress the frame.
+
+        The method is to compress the frame to protect the privacy of the frame.
```
Comment (Hulongyin) [inline cnap/core/frame.py]: Typo?

File: cnap/core/frame.py [hunk 14]
```diff
@@ -58,6 +58,46 @@ def decrypt(self) -> None:
...
+
+        Raises:
+            NotImplementedError: If the subclasses don't implement the method.
+        """
+        raise NotImplementedError("Subclasses should implement compress() method.")
+
+    def decompress(self, src: bytes) -> bytes:
+        """Decompress the frame.
+        
+        The method is to decompress the frame to protect the privacy of the frame.
```
Comment (Hulongyin) [inline cnap/core/frame.py]: Typo?

File: cnap/core/qat_zip.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
+"""A QAT params module.
+
+This module provides the relevant parameters required to call QAT.
+
+Classes:
+    QzDirection: An enum for the type of the Huffman header supported by QATzip.
+    QzPollingMode: An enum for the type of polling mode supported by QATzip.
```
Comment (Hulongyin) [inline cnap/core/qat_zip.py]: The description of this module is the same as `qat_params.py`

File: cnap/core/frame.py [hunk 16]
```diff
@@ -58,6 +58,46 @@ def decrypt(self) -> None:
         """
         raise NotImplementedError("Subclasses should implement decrypt method.")
 
+class FrameZipBase(ABC):
```
Comment (Hulongyin) [inline cnap/core/frame.py]: The description of this class should also be added to the docstring of this module.

Comment (Yanbo0101) [reply cnap/core/frame.py]: Fix all. Thanks!

**Comment by dongx1x** [comment]
@Yanbo0101 How about changing `qat_zip.py` to `qat.py` and `QATFrameZip` to `QAT`? then it can be used for more scenarios and add more features like encryption or decryption.

**Comment by Yanbo0101** [comment]
> @Yanbo0101 How about changing `qat_zip.py` to `qat.py` and `QATFrameZip` to `QAT`? then it can be used for more scenarios and add more features like encryption or decryption.

I have made the modifications and submitted them.  For encryption and decryption, initializing QAT and establishing a session are somewhat different from QATzip, so I have retained class QATzip. Could you please review again？Thanks!

File: cnap/core/qat_params.py [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    """An enum for the type of QzipHuffmanHdr.
+    
+    This enum defines the types of the Huffman header supported by QATzip.
+    """
+
+    QZ_DYNAMIC_HDR = ctypes.c_uint(0)
+    QZ_STATIC_HDR = ctypes.c_uint(1)
+
+
+# pylint: disable=R0903
```
Comment (leyao-daily) [inline cnap/core/qat_params.py]: Can you remove the comment to disable pylint check, we may configure the pylint check based on the information if needed.

File: cnap/core/qat.py [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+        if ret != 0:
+            LOG.error("Failed to decompress: %s.", ret)
+            raise RuntimeError(f"Decompression failed with status: {ret}.")
+
+        end_time = time.time()
+        time_cost = end_time - start_time
+
+        LOG.debug("Decompress %d bytes in %d bytes, cost %fs", len(src), dst_len.value, time_cost)
+        return bytes(dst_bytes_array.raw[:dst_len.value])
+
```
Comment (leyao-daily) [inline cnap/core/qat.py]: No newlint at the end?

File: cnap/core/qat_params.py (hunk 17 above)
Comment (Yanbo0101) [reply cnap/core/qat_params.py]: Can I fix them like line 30 in cnap/userv/streaming.py(# pylint: disable=too-many-instance-attributes) or just delete them?

Comment (leyao-daily) [reply cnap/core/qat_params.py]: I will update the pylint configuration to pass the check, so just delete it.

Comment (Yanbo0101) [reply cnap/core/qat_params.py]: ok, thanks!

File: cnap/core/qat.py (hunk 18 above)
Comment (Yanbo0101) [reply cnap/core/qat.py]: Fix all. Thanks!

**Review by Hulongyin — APPROVED** [review]
LGTM. Thanks

File: cnap/core/qat.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+    def comp_lvl(self) -> int:
+        """int: The QATzip compression level."""
+        return self._comp_lvl
+
+    @property
+    def algorithm(self) -> int:
+        """int: The QATzip compression/decompression algorithm."""
+        return self._algorithm
+
+    def init_session(self) :
```
Comment (leyao-daily) [inline cnap/core/qat.py]: Remove the space and maybe add the return hints like close_session

File: cnap/core/qat.py [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+        Raises:
+            RuntimeError: If any errors during the initialization of QAT. 
+        """
+        ret = qziplib.qzInit(ctypes.byref(self._session), self.SW_BACKUP)
+        if ret != 0:
+            LOG.error("Failed to initialize the QATzip session: %s", ret)
+            raise RuntimeError(f"Initialize QATzip session failed with status:{ret}.")
+        LOG.info("Initialize QAT hardware successful.")
+
+    def set_session(self) :
```
Comment (leyao-daily) [inline cnap/core/qat.py]: Same.

File: cnap/core/qat.py [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+        """
+        ret = qziplib.qzSetupSessionDeflate(
+            ctypes.byref(self._session),
+            ctypes.byref(self._deflate_params))
+        if ret != 0:
+            LOG.error("Failed to setup the QATzip session: %s.", ret)
+            raise RuntimeError(f"Set QATzip session failed with status: {ret}.")
+        LOG.info("Set up QAT session successful.")
+
+    def teardown_session(self) :
```
Comment (leyao-daily) [inline cnap/core/qat.py]: Same.

File: cnap/core/qat.py [hunk 22]
```diff
@@ -0,0 +1,231 @@
+"""A QAT module.
```
Comment (leyao-daily) [inline cnap/core/qat.py:1]: And do we need add the method or machnism to ensure that all session will be released (__del__), or just called by users (eg, close session). What do you think about it.

Comment (Yanbo0101) [reply cnap/core/qat.py:1]: Fix all. Thanks!

**Review by leyao-daily — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**